{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# default_exp diffquant_utils"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "import os\n",
    "import pathlib\n",
    "if \"__file__\" in globals():#only run in the translated python file, as __file__ is not defined with ipython\n",
    "    INTABLE_CONFIG = os.path.join(pathlib.Path(__file__).parent.absolute(), \"configs\", \"intable_config.yaml\") #the yaml config is located one directory below the python library files\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Utility Functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Reformatting and transformation functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "\n",
    "def get_samples_used_from_samplemap_file(samplemap_file, cond1, cond2):\n",
    "    samplemap_df = load_samplemap(samplemap_file)\n",
    "    return get_samples_used_from_samplemap_df(samplemap_df, cond1, cond2)\n",
    "\n",
    "\n",
    "def get_samples_used_from_samplemap_df(samplemap_df, cond1, cond2):\n",
    "    samples_c1 = samplemap_df[[cond1 == x for x in samplemap_df[\"condition\"]]][\"sample\"] #subset the df to the condition\n",
    "    samples_c2 = samplemap_df[[cond2 == x for x in samplemap_df[\"condition\"]]][\"sample\"]\n",
    "    return list(samples_c1), list(samples_c2)\n",
    "\n",
    "def get_all_samples_from_samplemap_df(samplemap_df):\n",
    "    return list(samplemap_df[\"sample\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "import pandas as pd\n",
    "\n",
    "def get_samplenames_from_input_df(data):\n",
    "    \"\"\"extracts the names of the samples of the AQ input dataframe\"\"\"\n",
    "    names = list(data.columns)\n",
    "    names.remove('protein')\n",
    "    names.remove('ion')\n",
    "    return names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "import numpy as np\n",
    "def filter_df_to_minrep(quant_df_wideformat, samples_c1, samples_c2, minrep):\n",
    "    \"\"\"filters dataframe in alphaquant format such that each column has a minimum number of replicates\n",
    "    \"\"\"\n",
    "    quant_df_wideformat = quant_df_wideformat.replace(0, np.nan)\n",
    "    df_c1_minrep = quant_df_wideformat[samples_c1].dropna(thresh = minrep, axis = 0)\n",
    "    df_c2_minrep = quant_df_wideformat[samples_c2].dropna(thresh = minrep, axis = 0)\n",
    "    idxs_both = df_c1_minrep.index.intersection(df_c2_minrep.index)\n",
    "    quant_df_reduced = quant_df_wideformat.iloc[idxs_both].reset_index()\n",
    "    return quant_df_reduced\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "def get_condpairname(condpair):\n",
    "    return f\"{condpair[0]}_VS_{condpair[1]}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "\n",
    "def get_quality_score_column(acquisition_info_df):\n",
    "    if \"FG.ShapeQualityScore\" in acquisition_info_df.columns:\n",
    "        param = \"FG.ShapeQualityScore\"\n",
    "    elif \"Quantity.Quality\" in acquisition_info_df.columns:\n",
    "        param = \"Quantity.Quality\"\n",
    "    return param"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "import os\n",
    "\n",
    "def make_dir_w_existcheck(dir):\n",
    "    if not os.path.exists(dir):\n",
    "        os.makedirs(dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "import os\n",
    "def get_results_plot_dir_condpair(results_dir, condpair):\n",
    "    results_dir_plots = f\"{results_dir}/{condpair}_plots\"\n",
    "    make_dir_w_existcheck(results_dir_plots)\n",
    "    return results_dir_plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "def get_middle_elem(sorted_list):\n",
    "    nvals = len(sorted_list)\n",
    "    if nvals==1:\n",
    "        return sorted_list[0]\n",
    "    middle_idx = nvals//2\n",
    "    if nvals%2==1:\n",
    "        return sorted_list[middle_idx]\n",
    "    return 0.5* (sorted_list[middle_idx] + sorted_list[middle_idx-1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "import numpy as np\n",
    "def get_nonna_array(array_w_nas):\n",
    "    res = []\n",
    "    isnan_arr = np.isnan(array_w_nas)\n",
    "\n",
    "    for idx in range(len(array_w_nas)):\n",
    "        sub_res = []\n",
    "        sub_array = array_w_nas[idx]\n",
    "        na_array = isnan_arr[idx]\n",
    "        for idx2 in range(len(sub_array)):\n",
    "            if not na_array[idx2]:\n",
    "               sub_res.append(sub_array[idx2])\n",
    "        res.append(np.array(sub_res))\n",
    "    return np.array(res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "import numpy as np\n",
    "def get_non_nas_from_pd_df(df):\n",
    "    return {\n",
    "        pep_name: sub_vals[~np.isnan(sub_vals)] for pep_name, sub_vals in\n",
    "        zip( df.index.values, df.values)\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "import numpy as np\n",
    "def get_ionints_from_pd_df(df):\n",
    "    return {\n",
    "        pep_name: sub_vals for pep_name, sub_vals in\n",
    "        zip( df.index.values, df.values)\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "def invert_dictionary(my_map):\n",
    "    inv_map = {}\n",
    "    for k, v in my_map.items():\n",
    "        inv_map[v] = inv_map.get(v, []) + [k]\n",
    "    return inv_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "import statistics\n",
    "\n",
    "def get_z_from_p_empirical(p_emp,p2z):\n",
    "    p_rounded = np.format_float_scientific(p_emp, 1)\n",
    "    if p_rounded in p2z:\n",
    "        return p2z.get(p_rounded)\n",
    "    z = statistics.NormalDist().inv_cdf(float(p_rounded))\n",
    "    p2z[p_rounded] = z\n",
    "    return z"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "import anytree\n",
    "def get_levelnodes_from_nodeslist(nodeslist, level):\n",
    "    levelnodes = []\n",
    "    for node in nodeslist:\n",
    "        precursors = anytree.findall(node, filter_= lambda x : (x.type == level))\n",
    "        levelnodes.extend(precursors)\n",
    "    return levelnodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "def count_fraction_outliers_from_expected_fc(result_df, threshold, expected_log2fc):\n",
    "    num_outliers = sum([abs(x-expected_log2fc)> threshold for x in result_df[\"log2fc\"]])\n",
    "    fraction_outliers = num_outliers/len(result_df[\"log2fc\"])\n",
    "    print(f\"{round(fraction_outliers, 2)} outliers\")\n",
    "    return fraction_outliers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "import anytree\n",
    "\n",
    "def find_node_parent_at_level(node, level):\n",
    "    if node.type == level:\n",
    "        return node\n",
    "    while node.parent is not None:\n",
    "        node = node.parent\n",
    "        if node.type == level:\n",
    "            return node\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "from anytree import Node\n",
    "\n",
    "def test_find_node_parent_at_level():\n",
    "\n",
    "    udo = Node(\"Udo\", type = 'granddad')\n",
    "    marc = Node(\"Marc\", parent=udo, type = 'dad')\n",
    "    lian = Node(\"Lian\", parent=marc, type = 'base')\n",
    "    dan = Node(\"Dan\", parent=udo, type ='dad')\n",
    "    jet = Node(\"Jet\", parent=dan, type ='base')\n",
    "    jan = Node(\"Jan\", parent=dan, type ='base')\n",
    "    joe = Node(\"Joe\", parent=dan, type ='base')\n",
    "\n",
    "    assert find_node_parent_at_level(lian, 'dad').name == 'Marc'\n",
    "    assert find_node_parent_at_level(lian, 'granddad').name == 'Udo'\n",
    "    assert find_node_parent_at_level(jet, 'dad').name == 'Dan'\n",
    "\n",
    "test_find_node_parent_at_level()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "import anytree\n",
    "\n",
    "def check_if_node_is_included(node):\n",
    "    while node.type != \"gene\":\n",
    "        if node.cluster != 0:\n",
    "            return False\n",
    "        node = node.parent\n",
    "    \n",
    "    return True\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "from anytree import Node\n",
    "from anytree import RenderTree\n",
    "\n",
    "\n",
    "def test_check_if_node_is_included():\n",
    "\n",
    "    udo = Node(\"Udo\", type = 'gene', cluster =-1)\n",
    "    marc = Node(\"Marc\", parent=udo, type = 'dad', cluster = 0)\n",
    "    lian = Node(\"Lian\", parent=marc, type = 'base', cluster = 0)\n",
    "    dan = Node(\"Dan\", parent=udo, type ='dad', cluster = 1)\n",
    "    jet = Node(\"Jet\", parent=dan, type ='base', cluster = 0)\n",
    "    jan = Node(\"Jan\", parent=dan, type ='base', cluster = 0)\n",
    "    joe = Node(\"Joe\", parent=dan, type ='base', cluster = 0)\n",
    "    print(RenderTree(udo))\n",
    "    assert check_if_node_is_included(jet) == False\n",
    "    assert check_if_node_is_included(lian) == True\n",
    "    assert check_if_node_is_included(dan) == False\n",
    "    assert check_if_node_is_included(marc) == True\n",
    "\n",
    "test_check_if_node_is_included()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## I/O functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "def write_chunk_to_file(chunk, filepath ,write_header):\n",
    "    \"\"\"write chunk of pandas dataframe to a file\"\"\"\n",
    "    chunk.to_csv(filepath, header=write_header, mode='a', sep = \"\\t\", index = None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Documentation functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "import logging\n",
    "import os\n",
    "import sys\n",
    "\n",
    "\n",
    "def set_logger(\n",
    "    *,log_path = None,\n",
    "    log_file_name=\"\",\n",
    "    stream: bool = True,\n",
    "    log_level: int = logging.INFO,\n",
    "    overwrite: bool = False,\n",
    ") -> str:\n",
    "    \"\"\"Set the log stream and file.\n",
    "    All previously set handlers will be disabled with this command.\n",
    "    Parameters\n",
    "    ----------\n",
    "    log_file_name : str, None\n",
    "        The file name to where the log is written.\n",
    "        Folders are automatically created if needed.\n",
    "        This is relative to the current path. When an empty string is provided,\n",
    "        a log is written to the AlphaTims \"logs\" folder with the name\n",
    "        \"log_yymmddhhmmss\" (reversed timestamp year to seconds).\n",
    "        If None, no log file is saved.\n",
    "        Default is \"\".\n",
    "    stream : bool\n",
    "        If False, no log data is sent to stream.\n",
    "        If True, all logging can be tracked with stdout stream.\n",
    "        Default is True.\n",
    "    log_level : int\n",
    "        The logging level. Usable values are defined in Python's \"logging\"\n",
    "        module.\n",
    "        Default is logging.INFO.\n",
    "    overwrite : bool\n",
    "        If True, overwrite the log_file if one exists.\n",
    "        If False, append to this log file.\n",
    "        Default is False.\n",
    "    Returns\n",
    "    -------\n",
    "    : str\n",
    "        The file name to where the log is written.\n",
    "    \"\"\"\n",
    "    import time\n",
    "    global PROGRESS_CALLBACK\n",
    "    log_path = os.path.join(BASE_PATH, \"logs\")\n",
    "    root = logging.getLogger()\n",
    "    formatter = logging.Formatter(\n",
    "        '%(asctime)s> %(message)s', \"%Y-%m-%d %H:%M:%S\"\n",
    "    )\n",
    "    root.setLevel(log_level)\n",
    "    while root.hasHandlers():\n",
    "        root.removeHandler(root.handlers[0])\n",
    "    if stream:\n",
    "        stream_handler = logging.StreamHandler(sys.stdout)\n",
    "        stream_handler.setLevel(log_level)\n",
    "        stream_handler.setFormatter(formatter)\n",
    "        root.addHandler(stream_handler)\n",
    "    if log_file_name is not None:\n",
    "        if log_file_name == \"\":\n",
    "            if not os.path.exists(log_path):\n",
    "                os.makedirs(log_path)\n",
    "            log_file_name = log_path\n",
    "        log_file_name = os.path.abspath(log_file_name)\n",
    "        if os.path.isdir(log_file_name):\n",
    "            current_time = time.localtime()\n",
    "            current_time = \"\".join(\n",
    "                [\n",
    "                    f'{current_time.tm_year:04}',\n",
    "                    f'{current_time.tm_mon:02}',\n",
    "                    f'{current_time.tm_mday:02}',\n",
    "                    f'{current_time.tm_hour:02}',\n",
    "                    f'{current_time.tm_min:02}',\n",
    "                    f'{current_time.tm_sec:02}',\n",
    "                ]\n",
    "            )\n",
    "            log_file_name = os.path.join(\n",
    "                log_file_name,\n",
    "                f\"log_{current_time}.txt\"\n",
    "            )\n",
    "        directory = os.path.dirname(log_file_name)\n",
    "        if not os.path.exists(directory):\n",
    "            os.makedirs(directory)\n",
    "        if overwrite:\n",
    "            file_handler = logging.FileHandler(log_file_name, mode=\"w\")\n",
    "        else:\n",
    "            file_handler = logging.FileHandler(log_file_name, mode=\"a\")\n",
    "        file_handler.setLevel(log_level)\n",
    "        file_handler.setFormatter(formatter)\n",
    "        root.addHandler(file_handler)\n",
    "    return log_file_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "import yaml\n",
    "import os.path\n",
    "import pathlib\n",
    "import re\n",
    "\n",
    "def load_method_parameters(results_dir):\n",
    "    params_file = f\"{results_dir}/aq_parameters.yaml\"\n",
    "    return load_config(params_file)\n",
    "\n",
    "def store_method_parameters(local_vars_dict, results_dir):\n",
    "    method_params = get_methods_dict_from_local_vars(local_vars_dict)\n",
    "    add_ml_input_file_location(method_params)\n",
    "    params_file = f\"{results_dir}/aq_parameters.yaml\"\n",
    "    if os.path.exists(params_file):\n",
    "        previous_params = load_method_parameters(results_dir)\n",
    "        previous_params = {x:y for x, y in previous_params.items() if x not in method_params.keys()}\n",
    "        method_params.update(previous_params)\n",
    "    if not os.path.exists(f\"{results_dir}/\"):\n",
    "        os.makedirs(f\"{results_dir}/\")\n",
    "    save_dict_as_yaml(method_params, params_file)\n",
    "\n",
    "def save_dict_as_yaml(dict, file):\n",
    "    with open(file, 'w') as outfile:\n",
    "        yaml.dump(dict, outfile, default_flow_style=False)\n",
    "\n",
    "\n",
    "def get_methods_dict_from_local_vars(local_vars):\n",
    "    method_params = {}\n",
    "    for x in local_vars.keys():\n",
    "        if local_vars[x] is None:\n",
    "            continue\n",
    "        if ((\"_df\" not in x) and ('condpair' not in x) and ('sys'!=x) and ('runconfig' != x)):\n",
    "            if (\"input_file\" in x) or (\"results_dir\" in x):\n",
    "                method_params[x] = os.path.abspath(local_vars[x])\n",
    "            else:\n",
    "                method_params[x] = local_vars[x]\n",
    "    return method_params\n",
    "\n",
    "def add_ml_input_file_location(method_params):\n",
    "    input_file = method_params.get(\"input_file\")\n",
    "    if \".aq_reformat\" in input_file:\n",
    "        ml_input_file = get_path_to_unformatted_file(input_file)\n",
    "    else:\n",
    "        ml_input_file = input_file\n",
    "    method_params[\"ml_input_file\"] = ml_input_file\n",
    "\n",
    "def get_path_to_unformatted_file(input_file_name):\n",
    "    prefixes = input_file_name.split(\".\")[:-3]\n",
    "    cleaned_filename = \".\".join(prefixes)\n",
    "    return cleaned_filename\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Input Parsers\n",
    "The Alphaquant pipeline is run using a generic wide-table input format, as specified in the documentation. The following parsers convert long format tables as provided e.g. by Spectronaut or DIA-NN into this generic format. The configuration for the parsers is set by a yaml file."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Convert long format to wide format"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Parse .yaml file\n",
    "The relevant parameters for reading and reformatting the long table are stored in the \"intable_config.yaml\" file. The functions below are for reading and reformating the config info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "import yaml\n",
    "import itertools\n",
    "\n",
    "def get_relevant_columns(protein_cols, ion_cols, sample_ID, quant_ID, filter_dict):\n",
    "    filtcols = []\n",
    "    for filtconf in filter_dict.values():\n",
    "        filtcols.append(filtconf.get('param'))\n",
    "    relevant_cols = protein_cols + ion_cols + [sample_ID] + [quant_ID] + filtcols\n",
    "    relevant_cols = list(set(relevant_cols)) # to remove possible redudancies\n",
    "    return relevant_cols\n",
    "\n",
    "\n",
    "def get_relevant_columns_config_dict(config_typedict):\n",
    "    filtcols = []\n",
    "    dict_ioncols = []\n",
    "    for filtconf in config_typedict.get('filters', {}).values():\n",
    "        filtcols.append(filtconf.get('param'))\n",
    "\n",
    "    if 'ion_hierarchy' in config_typedict.keys():\n",
    "        for headr in config_typedict.get('ion_hierarchy').values():\n",
    "            ioncols = list(itertools.chain.from_iterable(headr.get(\"mapping\").values()))\n",
    "            dict_ioncols.extend(ioncols)\n",
    "    quantID = config_typedict.get(\"quant_ID\")\n",
    "    if type(quantID) ==type(\"string\"):\n",
    "        quant_ids = [config_typedict.get(\"quant_ID\")]\n",
    "    elif quantID == None:\n",
    "        quant_ids = []\n",
    "    else:\n",
    "        quant_ids = list(config_typedict.get(\"quant_ID\").values())\n",
    "    relevant_cols = config_typedict.get(\"protein_cols\") + config_typedict.get(\"ion_cols\", []) + [config_typedict.get(\"sample_ID\")] + quant_ids + filtcols + dict_ioncols\n",
    "    relevant_cols = list(set(relevant_cols)) # to remove possible redudancies\n",
    "    return relevant_cols\n",
    "\n",
    "\n",
    "\n",
    "def get_config_columns(config_dict):\n",
    "    protein_cols = config_dict.get(\"protein_cols\")\n",
    "    ion_cols = config_dict.get(\"ion_cols\")\n",
    "    sample_ID = config_dict.get(\"sample_ID\")\n",
    "    quant_ID = config_dict.get(\"quant_ID\")\n",
    "    filter_dict = config_dict.get(\"filters\", {})\n",
    "    relevant_cols = get_relevant_columns(protein_cols, ion_cols, sample_ID, quant_ID, filter_dict)\n",
    "    return relevant_cols, protein_cols, ion_cols, sample_ID, quant_ID, filter_dict\n",
    "\n",
    "\n",
    "def load_config(config_yaml):\n",
    "    stream = open(config_yaml, 'r')\n",
    "    config_all = yaml.safe_load(stream)\n",
    "    return config_all\n",
    "\n",
    "def get_type2relevant_cols(config_all):\n",
    "    type2relcols = {}\n",
    "    for type in config_all.keys():\n",
    "        config_typedict = config_all.get(type)\n",
    "        relevant_cols = get_relevant_columns_config_dict(config_typedict)\n",
    "        type2relcols[type] = relevant_cols\n",
    "    return type2relcols"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filter and reformat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "\n",
    "def filter_input(filter_dict, input):\n",
    "    for filtname,filterconf in filter_dict.items():\n",
    "        param = filterconf.get('param')\n",
    "        comparator = filterconf.get('comparator')\n",
    "        value = filterconf.get('value')\n",
    "\n",
    "        if comparator not in [\">\",\">=\", \"<\", \"<=\", \"==\", \"!=\"]:\n",
    "            raise TypeError(f\"cannot identify the filter comparator of {filtname} given in the longtable config yaml!\")\n",
    "\n",
    "        if comparator==\"==\":\n",
    "            input = input[input[param] ==value]\n",
    "            continue\n",
    "        try:\n",
    "            input = input.astype({f\"{param}\" : \"float\"})\n",
    "        except:\n",
    "            pass\n",
    "\n",
    "        if comparator==\">\":\n",
    "            input = input[input[param].astype(type(value)) >value]\n",
    "\n",
    "        if comparator==\">=\":\n",
    "            input = input[input[param].astype(type(value)) >=value]\n",
    "\n",
    "        if comparator==\"<\":\n",
    "            input = input[input[param].astype(type(value)) <value]\n",
    "\n",
    "        if comparator==\"<=\":\n",
    "            input = input[input[param].astype(type(value)) <=value]\n",
    "\n",
    "        if comparator==\"!=\":\n",
    "            input = input[input[param].astype(type(value)) !=value]\n",
    "\n",
    "    return input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "def merge_protein_and_ion_cols(input_df, config_dict):\n",
    "    protein_cols =  config_dict.get(\"protein_cols\")\n",
    "    ion_cols = config_dict.get(\"ion_cols\")\n",
    "    input_df['protein'] = input_df.loc[:, protein_cols].astype('string').sum(axis=1)\n",
    "    input_df['ion'] = input_df.loc[:, ion_cols].astype('string').sum(axis=1)\n",
    "    input_df = input_df.rename(columns = {config_dict.get('quant_ID') : \"quant_val\"})\n",
    "    return input_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "import copy\n",
    "def merge_protein_cols_and_ion_dict(input_df, config_dict):\n",
    "    \"\"\"[summary]\n",
    "\n",
    "    Args:\n",
    "        input_df ([pandas dataframe]): longtable containing peptide intensity data\n",
    "        confid_dict ([dict[String[]]]): nested dict containing the parse information. derived from yaml file\n",
    "\n",
    "    Returns:\n",
    "        pandas dataframe: longtable with newly assigned \"protein\" and \"ion\" columns\n",
    "    \"\"\"\n",
    "    protein_cols = config_dict.get(\"protein_cols\")\n",
    "    ion_hierarchy = config_dict.get(\"ion_hierarchy\")\n",
    "    splitcol2sep = config_dict.get('split_cols')\n",
    "    quant_id_dict = config_dict.get('quant_ID')\n",
    "\n",
    "    ion_dfs = []\n",
    "    input_df['protein'] = input_df.loc[:, protein_cols].astype('string').sum(axis=1)\n",
    "\n",
    "    input_df = input_df.drop(columns = [x for x in protein_cols if x!='protein'])\n",
    "    for hierarchy_type in ion_hierarchy.keys():\n",
    "        df_subset = input_df.copy()\n",
    "        ion_hierarchy_local = ion_hierarchy.get(hierarchy_type).get(\"order\")\n",
    "        ion_headers_merged, ion_headers_grouped = get_ionname_columns(ion_hierarchy.get(hierarchy_type).get(\"mapping\"), ion_hierarchy_local) #ion headers merged is just a helper to select all relevant rows, ionheaders grouped contains the sets of ionstrings to be merged into a list eg [[SEQ, MOD], [CH]]\n",
    "        quant_columns = get_quantitative_columns(df_subset, hierarchy_type, config_dict, ion_headers_merged)\n",
    "        headers = list(set(ion_headers_merged + quant_columns + ['protein']))\n",
    "        if \"sample_ID\" in config_dict.keys():\n",
    "            headers+=[config_dict.get(\"sample_ID\")]\n",
    "        df_subset = df_subset[headers].drop_duplicates()\n",
    "\n",
    "        if splitcol2sep is not None:\n",
    "            if quant_columns[0] in splitcol2sep.keys(): #in the case that quantitative values are stored grouped in one column (e.g. msiso1,msiso2,msiso3, etc.), reformat accordingly\n",
    "                df_subset = split_extend_df(df_subset, splitcol2sep)\n",
    "            ion_headers_grouped = adapt_headers_on_extended_df(ion_headers_grouped, splitcol2sep)\n",
    "\n",
    "        #df_subset = df_subset.set_index(quant_columns)\n",
    "\n",
    "        df_subset = add_merged_ionnames(df_subset, ion_hierarchy_local, ion_headers_grouped, quant_id_dict, hierarchy_type)\n",
    "        ion_dfs.append(df_subset)\n",
    "    input_df = pd.concat(ion_dfs, ignore_index=True)\n",
    "    return input_df\n",
    "\n",
    "\n",
    "def get_quantitative_columns(input_df, hierarchy_type, config_dict, ion_headers_merged):\n",
    "    naming_columns = ion_headers_merged + ['protein']\n",
    "    if config_dict.get(\"format\") == 'longtable':\n",
    "        quantcol = config_dict.get(\"quant_ID\").get(hierarchy_type)\n",
    "        return [quantcol]\n",
    "\n",
    "    if config_dict.get(\"format\") == 'widetable':\n",
    "        quantcolumn_candidates = [x for x in input_df.columns if x not in naming_columns]\n",
    "        if \"quant_prefix\" in config_dict.keys():\n",
    "            return [x for x in quantcolumn_candidates if x.startswith(config_dict.get(\"quant_prefix\"))] # in the case that the quantitative columns have a prefix (like \"Intensity \" in MQ peptides.txt), only columns with the prefix are filtered\n",
    "        else:\n",
    "            return quantcolumn_candidates #in this case, we assume that all non-ionname/proteinname columns are quantitative columns\n",
    "\n",
    "\n",
    "def get_ionname_columns(ion_dict, ion_hierarchy_local):\n",
    "    ion_headers_merged = []\n",
    "    ion_headers_grouped = []\n",
    "    for lvl in ion_hierarchy_local:\n",
    "        vals = ion_dict.get(lvl)\n",
    "        ion_headers_merged.extend(vals)\n",
    "        ion_headers_grouped.append(vals)\n",
    "    return ion_headers_merged, ion_headers_grouped\n",
    "\n",
    "\n",
    "def adapt_headers_on_extended_df(ion_headers_grouped, splitcol2sep):\n",
    "    #in the case that one column has been split, we need to designate the \"naming\" column\n",
    "    ion_headers_grouped_copy = copy.deepcopy(ion_headers_grouped)\n",
    "    for vals in ion_headers_grouped_copy:\n",
    "        if splitcol2sep is not None:\n",
    "            for idx in range(len(vals)):\n",
    "                if vals[idx] in splitcol2sep.keys():\n",
    "                    vals[idx] = vals[idx] + \"_idxs\"\n",
    "    return ion_headers_grouped_copy\n",
    "\n",
    "def split_extend_df(input_df, splitcol2sep, value_threshold=10):\n",
    "    \"\"\"reformats data that is stored in a condensed way in a single column. For example isotope1_intensity;isotope2_intensity etc. in Spectronaut\n",
    "\n",
    "    Args:\n",
    "        input_df ([type]): [description]\n",
    "        splitcol2sep ([type]): [description]\n",
    "        value_threshold([type]): [description]\n",
    "\n",
    "    Returns:\n",
    "        Pandas Dataframe: Pandas dataframe with the condensed items expanded to long format\n",
    "    \"\"\"\n",
    "    if splitcol2sep==None:\n",
    "        return input_df\n",
    "\n",
    "    for split_col, separator in splitcol2sep.items():\n",
    "        idx_name = f\"{split_col}_idxs\"\n",
    "        split_col_series = input_df[split_col].str.split(separator)\n",
    "        input_df = input_df.drop(columns = [split_col])\n",
    "\n",
    "        input_df[idx_name] = [list(range(len(x))) for x in split_col_series]\n",
    "        exploded_input = input_df.explode(idx_name)\n",
    "        exploded_split_col_series = split_col_series.explode()\n",
    "\n",
    "        exploded_input[split_col] = exploded_split_col_series.replace('', 0) #the column with the intensities has to come after to column with the idxs\n",
    "\n",
    "        exploded_input = exploded_input.astype({split_col: float})\n",
    "        exploded_input = exploded_input[exploded_input[split_col]>value_threshold]\n",
    "        #exploded_input = exploded_input.rename(columns = {'var1': split_col})\n",
    "    return exploded_input\n",
    "\n",
    "\n",
    "\n",
    "def add_merged_ionnames(df_subset, ion_hierarchy_local, ion_headers_grouped, quant_id_dict, hierarchy_type):\n",
    "    \"\"\"puts together the hierarchical ion names as a column in a given input dataframe\"\"\"\n",
    "    all_ion_headers = list(itertools.chain.from_iterable(ion_headers_grouped))\n",
    "    columns_to_index = [x for x in df_subset.columns if x not in all_ion_headers]\n",
    "    df_subset = df_subset.set_index(columns_to_index)\n",
    "\n",
    "    rows = df_subset[all_ion_headers].to_numpy()\n",
    "    ions = []\n",
    "\n",
    "    for row in rows: #iterate through dataframe\n",
    "        count = 0\n",
    "        ionstring = \"\"\n",
    "        for lvl_idx in range(len(ion_hierarchy_local)):\n",
    "            ionstring += f\"{ion_hierarchy_local[lvl_idx]}\"\n",
    "            for sublvl in ion_headers_grouped[lvl_idx]:\n",
    "                ionstring+= f\"_{row[count]}_\"\n",
    "                count+=1\n",
    "        ions.append(ionstring)\n",
    "    df_subset['ion'] = ions\n",
    "    df_subset = df_subset.reset_index()\n",
    "    if quant_id_dict!= None:\n",
    "        df_subset = df_subset.rename(columns = {quant_id_dict.get(hierarchy_type) : \"quant_val\"})\n",
    "    return df_subset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "import os.path\n",
    "def reformat_and_write_longtable_according_to_config(input_file, outfile_name, config_dict, sep = \"\\t\",decimal = \".\", enforce_largefile_processing = False, chunksize =1000_000):\n",
    "    \"\"\"Reshape a long format proteomics results table (e.g. Spectronaut or DIA-NN) to a wide format table.\n",
    "    :param file input_file: long format proteomic results table\n",
    "    :param string input_type: the configuration key stored in the config file (e.g. \"diann_precursor\")\n",
    "    \"\"\"\n",
    "    filesize = os.path.getsize(input_file)/(1024**3) #size in gigabyte\n",
    "    file_is_large = (filesize>10 and str(input_file).endswith(\".zip\")) or filesize>50 or enforce_largefile_processing\n",
    "\n",
    "    if file_is_large:\n",
    "        tmpfile_large = f\"{input_file}.tmp.longformat.columnfilt.tsv\" #only needed when file is large\n",
    "        #remove potential leftovers from previous processings\n",
    "        if os.path.exists(tmpfile_large):\n",
    "            os.remove(tmpfile_large)\n",
    "        if os.path.exists(outfile_name):\n",
    "            os.remove(outfile_name)\n",
    "    \n",
    "    relevant_cols = get_relevant_columns_config_dict(config_dict)\n",
    "    input_df_it = pd.read_csv(input_file, sep = sep, decimal=decimal, usecols = relevant_cols, encoding ='latin1', chunksize = chunksize)\n",
    "    input_df_list = []\n",
    "    header = True\n",
    "    for input_df_subset in input_df_it:\n",
    "        input_df_subset = adapt_subtable(input_df_subset, config_dict)\n",
    "        if file_is_large:\n",
    "            write_chunk_to_file(input_df_subset,tmpfile_large, header)\n",
    "        else:\n",
    "            input_df_list.append(input_df_subset)\n",
    "        header = False\n",
    "        \n",
    "    if file_is_large:\n",
    "        process_with_dask(tmpfile_columnfilt=tmpfile_large , outfile_name = outfile_name, config_dict=config_dict)\n",
    "    else:\n",
    "        input_df = pd.concat(input_df_list)\n",
    "        input_reshaped = reshape_input_df(input_df, config_dict)\n",
    "        input_reshaped.to_csv(outfile_name, sep = \"\\t\", index = None)\n",
    "    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "import dask.dataframe as dd\n",
    "import pandas as pd\n",
    "import glob\n",
    "import os\n",
    "import shutil \n",
    "\n",
    "def adapt_subtable(input_df_subset, config_dict):\n",
    "    input_df_subset = filter_input(config_dict.get(\"filters\", {}), input_df_subset)\n",
    "    if \"ion_hierarchy\" in config_dict.keys():\n",
    "        return merge_protein_cols_and_ion_dict(input_df_subset, config_dict)\n",
    "    else:\n",
    "        return merge_protein_and_ion_cols(input_df_subset, config_dict)\n",
    "\n",
    "\n",
    "def reshape_input_df(input_df, config_dict):\n",
    "    input_df = input_df.astype({'quant_val': 'float'})\n",
    "    input_reshaped = pd.pivot_table(input_df, index = ['protein', 'ion'], columns = config_dict.get(\"sample_ID\"), values = 'quant_val', fill_value=0)\n",
    "    if input_reshaped.iloc[:,0].replace(0, np.nan).median() <100: #when values are small, rescale by a constant factor to prevent rounding errors in the subsequent aq analyses\n",
    "        input_reshaped = input_reshaped *10000\n",
    "\n",
    "    input_reshaped = input_reshaped.reset_index()\n",
    "    return input_reshaped\n",
    "\n",
    "\n",
    "def process_with_dask(*, tmpfile_columnfilt, outfile_name, config_dict):\n",
    "    df = dd.read_csv(tmpfile_columnfilt, sep = \"\\t\")\n",
    "    allcols = df[config_dict.get(\"sample_ID\")].drop_duplicates().compute()\n",
    "    allcols = ['protein', 'ion'] + sorted(allcols)\n",
    "    df = df.set_index('protein')\n",
    "    sorted_filedir = f\"{tmpfile_columnfilt}_sorted\"\n",
    "    df.to_csv(sorted_filedir, sep = \"\\t\")\n",
    "    #now the files are sorted and can be pivoted chunkwise (multiindex pivoting at the moment not possible in dask)\n",
    "    files_dask = glob.glob(f\"{sorted_filedir}/*part\")\n",
    "    header = True\n",
    "    for file in files_dask:\n",
    "        input_df = pd.read_csv(file, sep = \"\\t\")\n",
    "        if len(input_df.index) <2:\n",
    "            continue\n",
    "        input_reshaped = reshape_input_df(input_df, config_dict)\n",
    "        input_reshaped = sort_and_add_columns(input_reshaped, allcols)\n",
    "        write_chunk_to_file(input_reshaped, outfile_name, header)\n",
    "        header = False\n",
    "    os.remove(tmpfile_columnfilt)\n",
    "    shutil.rmtree(sorted_filedir)\n",
    "\n",
    "\n",
    "def sort_and_add_columns(input_reshaped, allcols):\n",
    "    missing_cols = set(allcols) - set(input_reshaped.columns)\n",
    "    input_reshaped[list(missing_cols)] = 0\n",
    "    input_reshaped = input_reshaped[allcols]\n",
    "    return input_reshaped\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "def reformat_and_write_wideformat_table(peptides_tsv, outfile_name, config_dict):\n",
    "    input_df = pd.read_csv(peptides_tsv,sep=\"\\t\", encoding ='latin1')\n",
    "    filter_dict = config_dict.get(\"filters\")\n",
    "    protein_cols = config_dict.get(\"protein_cols\")\n",
    "    ion_cols = config_dict.get(\"ion_cols\")\n",
    "    input_df = filter_input(filter_dict, input_df)\n",
    "    #input_df = merge_protein_and_ion_cols(input_df, config_dict)\n",
    "    input_df = merge_protein_cols_and_ion_dict(input_df, config_dict)\n",
    "    if 'quant_prefix' in config_dict.keys():\n",
    "        quant_prefix = config_dict.get('quant_prefix')\n",
    "        headers = ['protein', 'ion'] + list(filter(lambda x: x.startswith(quant_prefix), input_df.columns))\n",
    "        input_df = input_df[headers]\n",
    "        input_df = input_df.rename(columns = lambda x : x.replace(quant_prefix, \"\"))\n",
    "\n",
    "    input_df = input_df.reset_index()\n",
    "    \n",
    "    input_df.to_csv(outfile_name, sep = '\\t', index = None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def read_mq_peptides_table(peptides_tsv, pepheader = \"Sequence\", protheader = \"Leading razor protein\"):\n",
    "    peps = pd.read_csv(peptides_tsv,sep=\"\\t\", encoding ='latin1')\n",
    "    peps = peps[peps[\"Reverse\"] != \"+\"]\n",
    "    peps = peps[peps[\"Potential contaminant\"] != \"+\"]\n",
    "    if pepheader != None:\n",
    "        peps = peps.rename(columns = {pepheader : \"ion\"})\n",
    "    if protheader != None:\n",
    "        peps = peps.rename(columns = {protheader: \"protein\"})\n",
    "    headers = ['protein', 'ion'] + list(filter(lambda x: x.startswith(\"Intensity \"), peps.columns))\n",
    "    peps = peps[headers]\n",
    "    peps = peps.rename(columns = lambda x : x.replace(\"Intensity \", \"\"))\n",
    "\n",
    "    return peps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "from anytree.importer import JsonImporter\n",
    "import os\n",
    "\n",
    "def read_condpair_tree(cond1, cond2, results_folder = os.path.join(\".\", \"results\")):\n",
    "    \"\"\"reads the merged and clustered iontree for a given condpair\"\"\"\n",
    "    condpairname = get_condpairname([cond1, cond2])\n",
    "    tree_file =os.path.join(results_folder, f\"{condpairname}.iontrees.json\")\n",
    "    if not os.path.isfile(tree_file):\n",
    "        return None\n",
    "    importer = JsonImporter()\n",
    "    filehandle = open(tree_file, 'r')\n",
    "    jsontree = importer.read(filehandle)\n",
    "    filehandle.close()\n",
    "    return jsontree"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## check for already processed files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "import os\n",
    "def check_for_processed_runs_in_results_folder(results_folder):\n",
    "    contained_condpairs = []\n",
    "    folder_files = os.listdir(results_folder)\n",
    "    result_files = list(filter(lambda x: \"results.tsv\" in x ,folder_files))\n",
    "    for result_file in result_files:\n",
    "        res_name = result_file.replace(\".results.tsv\", \"\")\n",
    "        if ((f\"{res_name}.normed.tsv\" in folder_files) & (f\"{res_name}.results.ions.tsv\" in folder_files)):\n",
    "            contained_condpairs.append(res_name)\n",
    "    return contained_condpairs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Wrapper functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "import pandas as pd\n",
    "import os\n",
    "import pathlib\n",
    "\n",
    "def import_data(input_file, input_type_to_use = None, samples_subset = None, results_dir = None):\n",
    "    \"\"\"\n",
    "    Function to import peptide level data. Depending on available columns in the provided file,\n",
    "    the function identifies the type of input used (e.g. Spectronaut, MaxQuant, DIA-NN), reformats if necessary\n",
    "    and returns a generic wide-format dataframe\n",
    "    :param file input_file: quantified peptide/ion -level data\n",
    "    :param file results_folder: the folder where the AlphaQuant outputs are stored\n",
    "    \"\"\"\n",
    "\n",
    "    samples_subset = add_ion_protein_headers_if_applicable(samples_subset)\n",
    "    if \"aq_reformat\" in input_file:\n",
    "        file_to_read = input_file\n",
    "    else:\n",
    "        file_to_read = reformat_and_save_input_file(input_file=input_file, input_type_to_use=input_type_to_use)\n",
    "    \n",
    "    input_reshaped = pd.read_csv(file_to_read, sep = \"\\t\", encoding = 'latin1', usecols=samples_subset)\n",
    "    return input_reshaped\n",
    "\n",
    "\n",
    "def reformat_and_save_input_file(input_file, input_type_to_use = None):\n",
    "    \n",
    "    input_type, config_dict_for_type, sep = get_input_type_and_config_dict(input_file, input_type_to_use)\n",
    "    print(f\"using input type {input_type}\")\n",
    "    format = config_dict_for_type.get('format')\n",
    "    outfile_name = f\"{input_file}.{input_type}.aq_reformat.tsv\"\n",
    "\n",
    "    if format == \"longtable\":\n",
    "        reformat_and_write_longtable_according_to_config(input_file, outfile_name,config_dict_for_type, sep = sep)\n",
    "    elif format == \"widetable\":\n",
    "        reformat_and_write_wideformat_table(input_file, outfile_name, config_dict_for_type)\n",
    "    else:\n",
    "        raise Exception('Format not recognized!')\n",
    "    return outfile_name\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def add_ion_protein_headers_if_applicable(samples_subset):\n",
    "    if samples_subset is not None:\n",
    "        return samples_subset + [\"ion\", \"protein\"]\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "import pandas as pd\n",
    "import os.path\n",
    "import pathlib\n",
    "\n",
    "def get_input_type_and_config_dict(input_file, input_type_to_use = None):\n",
    "    #parse the type of input (e.g. Spectronaut Fragion+MS1Iso) out of the input file\n",
    "\n",
    "\n",
    "    config_dict = load_config(INTABLE_CONFIG)\n",
    "    type2relevant_columns = get_type2relevant_cols(config_dict)\n",
    "\n",
    "    if \"aq_reformat.tsv\" in input_file:\n",
    "        input_file = get_original_file_from_aq_reformat(input_file)\n",
    "\n",
    "    filename = str(input_file)\n",
    "    if '.csv' in filename:\n",
    "        sep=','\n",
    "    if '.tsv' in filename:\n",
    "        sep='\\t'\n",
    "    if '.txt' in filename:\n",
    "        sep='\\t'\n",
    "\n",
    "    if 'sep' not in locals():\n",
    "        raise TypeError(f\"neither of the file extensions (.tsv, .csv, .txt) detected for file {input_file}! Your filename has to contain one of these extensions. Please modify your file name accordingly.\")\n",
    "\n",
    "\n",
    "\n",
    "    uploaded_data_columns = set(pd.read_csv(input_file, sep=sep, nrows=1, encoding ='latin1').columns)\n",
    "\n",
    "    for input_type in type2relevant_columns.keys():\n",
    "        if (input_type_to_use is not None) and (input_type!=input_type_to_use):\n",
    "            continue\n",
    "        relevant_columns = type2relevant_columns.get(input_type)\n",
    "        relevant_columns = [x for x in relevant_columns if x] #filter None values\n",
    "        if set(relevant_columns).issubset(uploaded_data_columns):\n",
    "            config_dict_type =  config_dict.get(input_type)\n",
    "            return input_type, config_dict_type, sep\n",
    "    raise TypeError(\"format not specified in intable_config.yaml!\")\n",
    "\n",
    "import re\n",
    "def get_original_file_from_aq_reformat(input_file):\n",
    "    matched = re.match(\"(.*)(\\..*\\.)(aq_reformat\\.tsv)\",input_file)\n",
    "    return matched.group(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "\n",
    "def test_get_original_file_from_aq_reformat():\n",
    "    assert get_original_file_from_aq_reformat(\"yeast_report_fastafiltered.tsv.some.oth.erstuff.spectronaut_fragion_isotopes.aq_reformat.tsv\") == \"yeast_report_fastafiltered.tsv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "def import_config_dict():\n",
    "    config_dict = load_config(INTABLE_CONFIG)\n",
    "    return config_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "def load_samplemap(samplemap_file):\n",
    "    file_ext = os.path.splitext(samplemap_file)[-1]\n",
    "    if file_ext=='.csv':\n",
    "        sep=','\n",
    "    if (file_ext=='.tsv') | (file_ext=='.txt'):\n",
    "        sep='\\t'\n",
    "\n",
    "    if 'sep' not in locals():\n",
    "        print(f\"neither of the file extensions (.tsv, .csv, .txt) detected for file {samplemap_file}! Trying with tab separation. In the case that it fails, please add the appropriate extension to your file name.\")\n",
    "        sep = \"\\t\"\n",
    "\n",
    "    return pd.read_csv(samplemap_file, sep = sep, encoding ='latin1', dtype='str')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "def prepare_loaded_tables(data_df, samplemap_df):\n",
    "    \"\"\"\n",
    "    Integrates information from the peptide/ion data and the samplemap, selects the relevant columns and log2 transforms intensities.\n",
    "    \"\"\"\n",
    "    samplemap_df = samplemap_df[samplemap_df[\"condition\"]!=\"\"] #remove rows that have no condition entry\n",
    "    filtvec_not_in_data = [(x in data_df.columns) for x in samplemap_df[\"sample\"]] #remove samples that are not in the dataframe\n",
    "    samplemap_df = samplemap_df[filtvec_not_in_data]\n",
    "    headers = ['protein'] + samplemap_df[\"sample\"].to_list()\n",
    "    data_df = data_df.set_index(\"ion\")\n",
    "    for sample in samplemap_df[\"sample\"]:\n",
    "        data_df[sample] = np.log2(data_df[sample].replace(0, np.nan))\n",
    "    return data_df[headers], samplemap_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "\n",
    "\n",
    "#export\n",
    "class LongTableReformater():\n",
    "    \"\"\"Generic class to reformat tabular files in chunks. For the specific cases you can inherit the class and specify reformat and iterate function\n",
    "    \"\"\"\n",
    "    def __init__(self, input_file):\n",
    "        self._input_file = input_file\n",
    "        self._reformatting_function = None\n",
    "        self._iterator_function = self.__initialize_df_iterator__\n",
    "        self._concat_list = []\n",
    "\n",
    "    def reformat_and_load_acquisition_data_frame(self):\n",
    "\n",
    "        input_df_it = self._iterator_function()\n",
    "        \n",
    "        input_df_list = []\n",
    "        for input_df_subset in input_df_it:\n",
    "            input_df_subset = self._reformatting_function(input_df_subset)\n",
    "            input_df_list.append(input_df_subset)\n",
    "        input_df = pd.concat(input_df_list)\n",
    "        \n",
    "        return input_df\n",
    "\n",
    "    def reformat_and_save_acquisition_data_frame(self, output_file):\n",
    "        \n",
    "        input_df_it = self._iterator_function()\n",
    "        write_header = True\n",
    "        \n",
    "        for input_df_subset in input_df_it:\n",
    "            input_df_subset = self._reformatting_function(input_df_subset)\n",
    "            self.__write_reformatted_df_to_file__(input_df_subset, output_file, write_header)\n",
    "            write_header = False\n",
    "\n",
    "    def __initialize_df_iterator__(self):\n",
    "        return pd.read_csv(self._input_file, sep = \"\\t\", encoding ='latin1', chunksize=1000000)\n",
    "    \n",
    "    @staticmethod\n",
    "    def __write_reformatted_df_to_file__(reformatted_df, filepath ,write_header):\n",
    "        reformatted_df.to_csv(filepath, header=write_header, mode='a', sep = \"\\t\", index = None)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "\n",
    "import os\n",
    "\n",
    "class AcquisitionTableHandler():\n",
    "    def __init__(self, results_dir, samples):\n",
    "        self._samples = samples\n",
    "        self._table_infos = AcquisitionTableInfo(results_dir=results_dir)\n",
    "        self._header_infos = AcquisitionTableHeaders(self._table_infos)\n",
    "    \n",
    "    def get_acquisition_info_df(self):\n",
    "        return self.__get_reformated_df__()\n",
    "\n",
    "    def save_dataframe_as_new_acquisition_dataframe(self):\n",
    "        self._output_paths = AcquisitionTableOutputPaths(self._table_infos)\n",
    "        self.__remove_possible_pre_existing_ml_table__(self._output_paths.output_file_name)\n",
    "        df_reformater = AcquisitionTableReformater(table_infos = self._table_infos, header_infos=self._header_infos, samples = self._samples, dataframe_already_preformated=False)\n",
    "        df_reformater.reformat_and_save_acquisition_data_frame(self._output_paths.output_file_name)\n",
    "\n",
    "    def update_ml_file_location_in_method_parameters_yaml(self):\n",
    "        method_params = load_method_parameters(self._table_infos._results_dir)\n",
    "        if self._output_paths == None:\n",
    "            raise Exception(\"output paths not initialized! This could be because no dataframe was saved before\")\n",
    "        method_params[self._output_paths.ml_file_accession_in_yaml] = self._output_paths.output_file_name\n",
    "        save_dict_as_yaml(method_params, self._output_paths.method_parameters_yaml_path)\n",
    "    \n",
    "    def __get_reformated_df__(self):\n",
    "        df_reformater = AcquisitionTableReformater(table_infos = self._table_infos, header_infos=self._header_infos, samples = self._samples, dataframe_already_preformated=True)\n",
    "        df = df_reformater.reformat_and_load_acquisition_data_frame()\n",
    "        return df.convert_dtypes()\n",
    "    \n",
    "    @staticmethod\n",
    "    def __remove_possible_pre_existing_ml_table__(output_file_name):\n",
    "        if os.path.exists(output_file_name):\n",
    "            os.remove(output_file_name)\n",
    "            print(f\"removed pre existing {output_file_name}\")\n",
    "\n",
    "\n",
    "class AcquisitionTableInfo():\n",
    "    def __init__(self, results_dir, sep = \"\\t\", decimal = \".\"):\n",
    "        self._results_dir = results_dir\n",
    "        self._sep = sep\n",
    "        self._decimal = decimal\n",
    "        self._input_file = self.__get_input_file__()\n",
    "        self._file_ending_of_formatted_table = \".ml_info_table.tsv\"\n",
    "        self.already_formatted =  self.__check_if_input_file_is_already_formatted__()\n",
    "        self._input_type, self._config_dict = self.__get_input_type_and_config_dict__()\n",
    "        self._sample_column = self.__get_sample_column__()\n",
    "        self.last_ion_level_to_use = self.__get_last_ion_level_to_use__()\n",
    "\n",
    "    def __get_input_file__(self):\n",
    "        method_params_dict = load_method_parameters(self._results_dir)\n",
    "        return method_params_dict.get('ml_input_file')\n",
    "\n",
    "    def __check_if_input_file_is_already_formatted__(self):\n",
    "        if self._file_ending_of_formatted_table in self._input_file:\n",
    "            return True\n",
    "        else:\n",
    "            return False\n",
    "\n",
    "    def __get_input_type_and_config_dict__(self):\n",
    "        if self.already_formatted:\n",
    "            original_file = self.__get_location_of_original_file__()\n",
    "        else:\n",
    "            original_file = self._input_file\n",
    "        input_type, config_dict, _ = get_input_type_and_config_dict(original_file)\n",
    "        return input_type, config_dict\n",
    "    \n",
    "    def __get_location_of_original_file__(self):\n",
    "        return self._input_file.replace(self._file_ending_of_formatted_table, \"\")\n",
    "    \n",
    "    def __get_sample_column__(self):\n",
    "        return self._config_dict.get(\"sample_ID\")\n",
    "        \n",
    "    def __get_last_ion_level_to_use__(self):\n",
    "        return self._config_dict[\"ml_level\"]\n",
    "\n",
    "\n",
    "class AcquisitionTableHeaders():\n",
    "    def __init__(self, acquisition_table_info):\n",
    "\n",
    "        self._table_info = acquisition_table_info\n",
    "\n",
    "        self._ion_hierarchy = self.__get_ordered_ion_hierarchy__()\n",
    "        self._included_levelnames = self.__get_included_levelnames__()\n",
    "        self._ion_headers_grouped = self.__get_ion_headers_grouped__()\n",
    "        self._ion_headers = self.__get_ion_headers__()\n",
    "        self._numeric_headers = self.__get_numeric_headers__()\n",
    "        self._relevant_headers = self.__get_relevant_headers__()\n",
    "    \n",
    "    def __get_ordered_ion_hierarchy__(self):\n",
    "        ion_hierarchy = self._table_info._config_dict.get(\"ion_hierarchy\")\n",
    "        hier_key = 'fragion' if 'fragion' in ion_hierarchy.keys() else list(ion_hierarchy.keys())[0]\n",
    "        ion_hierarchy_on_chosen_key = ion_hierarchy.get(hier_key)\n",
    "        return ion_hierarchy_on_chosen_key\n",
    "\n",
    "    def __get_included_levelnames__(self):\n",
    "        levelnames = self.__get_all_levelnames__(self._ion_hierarchy)\n",
    "        last_ionlevel_idx = levelnames.index(self._table_info.last_ion_level_to_use)\n",
    "        return levelnames[:last_ionlevel_idx+1]\n",
    "    \n",
    "    @staticmethod\n",
    "    def __get_all_levelnames__(ion_hierarchy):\n",
    "        return  ion_hierarchy.get('order')\n",
    "\n",
    "    def __get_ion_headers_grouped__(self):\n",
    "        mapping_dict = self.__get_levelname_mapping_dict(self._ion_hierarchy)\n",
    "        return [mapping_dict.get(x) for x in self._included_levelnames]#on each level there can be multiple names, so it is a list of lists\n",
    "\n",
    "    @staticmethod\n",
    "    def __get_levelname_mapping_dict(ion_hierarchy):\n",
    "        return ion_hierarchy.get('mapping')\n",
    "    \n",
    "    def __get_ion_headers__(self):\n",
    "        return list(itertools.chain(*self._ion_headers_grouped))\n",
    "\n",
    "    \n",
    "    def __get_relevant_headers__(self):\n",
    "        relevant_headers = self._numeric_headers+self._ion_headers + [self._table_info._sample_column]\n",
    "        return self.__remove_possible_none_values_from_list__(relevant_headers)\n",
    "    \n",
    "    @staticmethod\n",
    "    def __remove_possible_none_values_from_list__(list):\n",
    "        return [x for x in list if x is not None]\n",
    "\n",
    "    def __get_numeric_headers__(self):\n",
    "        df_sample = pd.read_csv(self._table_info._input_file, sep = self._table_info._sep, decimal = self._table_info._decimal, encoding='latin1', nrows=3000) #sample 3000 rows from the df to assess the types of each row\n",
    "        df_sample = df_sample.replace({False: 0, True: 1})\n",
    "        numeric_headers =  list(df_sample.select_dtypes(include=np.number).columns)\n",
    "        numeric_headers = AcquisitionTableHeaderFilter().filter_numeric_headers_if_specified(input_type = self._table_info._input_type, numeric_headers = numeric_headers)\n",
    "        return numeric_headers\n",
    "\n",
    "\n",
    "class AcquisitionTableOutputPaths():\n",
    "    def __init__(self, table_info):\n",
    "        self._table_info = table_info\n",
    "        self.output_file_name = self.__get_output_file_name__()\n",
    "        self.method_parameters_yaml_path = self.__get_method_parameters_yaml_path__()\n",
    "        self.ml_file_accession_in_yaml = \"ml_input_file\"\n",
    "\n",
    "    def __get_output_file_name__(self):\n",
    "        old_file_name = self._table_info._input_file\n",
    "        new_file_name = old_file_name+self._table_info._file_ending_of_formatted_table\n",
    "        return new_file_name\n",
    "\n",
    "    def __get_method_parameters_yaml_path__(self):\n",
    "        return f\"{self._table_info._results_dir}/aq_parameters.yaml\"\n",
    "\n",
    "\n",
    "class AcquisitionTableReformater(LongTableReformater):\n",
    "    def __init__(self, table_infos, header_infos, samples, dataframe_already_preformated = False):\n",
    "        \n",
    "        LongTableReformater.__init__(self, table_infos._input_file)\n",
    "        self._table_infos = table_infos\n",
    "        self._header_infos = header_infos\n",
    "        self._samples = samples\n",
    "        self._dataframe_already_preformated = dataframe_already_preformated\n",
    "\n",
    "        #set the two functions that specify the explicit reformatting\n",
    "        self._reformatting_function = self.__reformatting_function__\n",
    "        self._iterator_function = self.__initialize_iterator_with_specified_columns__\n",
    "    \n",
    "    def __reformatting_function__(self, input_df_subset):\n",
    "        input_df_subset = input_df_subset.drop_duplicates()\n",
    "        input_df_subset = self.__filter_reformated_df_if_necessary__(input_df_subset)\n",
    "        if not self._dataframe_already_preformated:\n",
    "            input_df_subset = add_merged_ionnames(input_df_subset, self._header_infos._included_levelnames, self._header_infos._ion_headers_grouped, None, None)\n",
    "        return input_df_subset\n",
    "\n",
    "    def __filter_reformated_df_if_necessary__(self, reformatted_df):\n",
    "        if 'maxquant' in self._table_infos._input_type:\n",
    "            return reformatted_df\n",
    "        if 'spectronaut' in self._table_infos._input_type or 'diann' in self._table_infos._input_type:\n",
    "            return self.__filter_reformatted_dataframe_to_relevant_samples__(reformatted_df)\n",
    "\n",
    "    def __filter_reformatted_dataframe_to_relevant_samples__(self, input_df_subset):\n",
    "        return input_df_subset[[x in self._samples for x in input_df_subset[self._table_infos._sample_column]]]\n",
    "    \n",
    "    def __initialize_iterator_with_specified_columns__(self):\n",
    "        cols_to_use = self.__get_cols_to_use__()\n",
    "        return pd.read_csv(self._table_infos._input_file, sep = self._table_infos._sep, decimal=self._table_infos._decimal, usecols = cols_to_use, encoding ='latin1', chunksize=1000000)\n",
    "\n",
    "    def __get_cols_to_use__(self):\n",
    "        cols_to_use = self._header_infos._relevant_headers\n",
    "        if self._dataframe_already_preformated:\n",
    "            return cols_to_use+['ion']\n",
    "        else:\n",
    "            return cols_to_use\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "class AcquisitionTableHeaderFilter():\n",
    "    def __init__(self):\n",
    "        self._spectronaut_header_filter = lambda x : ((\"EG.\" in x) | (\"FG.\" in x)) and (\"Global\" not in x)\n",
    "        self._maxquant_header_filter = lambda x : (\"Intensity\" not in x) and (\"Experiment\" not in x)\n",
    "\n",
    "    def filter_numeric_headers_if_specified(self, input_type, numeric_headers):\n",
    "        if 'spectronaut' in input_type:\n",
    "            return [x for x in numeric_headers if self._spectronaut_header_filter(x)]\n",
    "        elif 'maxquant' in input_type:\n",
    "            return [x for x in numeric_headers if self._maxquant_header_filter(x)]\n",
    "        else:\n",
    "            return numeric_headers\n",
    "\n",
    "\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "\n",
    "def merge_acquisition_df_parameter_df(acquisition_df, parameter_df, groupby_merge_type = 'mean'):\n",
    "    \"\"\"acquisition df contains details on the acquisition, parameter df are the parameters derived from the tree\n",
    "    \"\"\"\n",
    "    merged_df = parameter_df.merge(acquisition_df, how = 'left', on = 'ion')\n",
    "    if groupby_merge_type == 'mean':\n",
    "        merged_df = merged_df.groupby('ion').mean().reset_index()\n",
    "    if groupby_merge_type == 'min':\n",
    "        merged_df = merged_df.groupby('ion').min().reset_index()\n",
    "    if groupby_merge_type == 'max':\n",
    "        merged_df = merged_df.groupby('ion').max().reset_index()\n",
    "    merged_df = merged_df.dropna(axis=1, how='all')\n",
    "    return merged_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Test input parsing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Compare the in-memory and out-of-memory longformat table loadings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "\n",
    "import os\n",
    "import pandas as pd\n",
    "import shutil\n",
    "import matplotlib_venn\n",
    "import alphaquant.diffquant_utils as aqutils\n",
    "\n",
    "print(os.path.abspath(\".\"))\n",
    "input_file = \"./test_data/unit_tests/input_table_formats/spectronaut.frgions.large.tsv\"\n",
    "outdir = \"./test_data/unit_tests/input_table_formats/loading_comparisons\"\n",
    "\n",
    "file_default = \"default_out.tsv\"\n",
    "file_dask_proc = \"dask_proc_out.tsv\"\n",
    "\n",
    "def test_table_loadings(input_file, outdir, file_default, file_dask_proc):\n",
    "    if not os.path.exists(outdir):\n",
    "        os.mkdir(outdir)\n",
    "        #os.chdir(outdir)\n",
    "\n",
    "    input_type, config_dict_for_type, sep = aqutils.get_input_type_and_config_dict(input_file)\n",
    "\n",
    "    aqutils.reformat_and_write_longtable_according_to_config(input_file,outfile_name=file_dask_proc,config_dict=config_dict_for_type, enforce_largefile_processing=True, chunksize=10_000)\n",
    "    aqutils.reformat_and_write_longtable_according_to_config(input_file,outfile_name=file_default,config_dict=config_dict_for_type, chunksize=10_000)\n",
    "\n",
    "    df_default = pd.read_csv(file_default, sep = \"\\t\")\n",
    "\n",
    "    df_dask_proc = pd.read_csv(file_dask_proc, sep = \"\\t\")\n",
    "\n",
    "\n",
    "    assert df_default.equals(df_dask_proc)\n",
    "\n",
    "    matplotlib_venn.venn2([set(df_default[\"ion\"]), set(df_dask_proc[\"ion\"])])\n",
    "\n",
    "    os.remove(file_default)\n",
    "    os.remove(file_dask_proc)\n",
    "    shutil.rmtree(outdir)\n",
    "\n",
    "\n",
    "test_table_loadings(input_file, outdir, file_default, file_dask_proc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "\n",
    "import alphaquant.diffquant_utils as aqutils\n",
    "import alphaquant.diff_analysis_manager as diffmgr\n",
    "import os\n",
    "import shutil\n",
    "outdir = \"./test_data/unit_tests/input_table_formats/\"\n",
    "\n",
    "tabledir =\"./test_data/unit_tests/input_table_formats/\"\n",
    "results_dir = \"./test_data/unit_tests/results/\"\n",
    "\n",
    "\n",
    "\n",
    "input_file = os.path.join(tabledir,\"diann.tsv\" )\n",
    "#input_file = os.path.join(tabledir, \"spectronaut.tsv\")\n",
    "#input_file = os.path.join(tabledir, \"spectronaut_frgion.tsv\")\n",
    "#input_file = os.path.join(tabledir, \"mq_peptides.txt\")\n",
    "\n",
    "samplemap_file = os.path.join(tabledir, \"samplemap.diann.tsv\")\n",
    "#samplemap_file = os.path.join(tabledir, \"samplemap.spectronaut.tsv\")\n",
    "#samplemap_file = os.path.join(tabledir, \"samplemap.spectronaut.frgions.tsv\")\n",
    "#samplemap_file = os.path.join(tabledir, \"samplemap.mq.tsv\")\n",
    "\n",
    "#import the input table once the input and the results folder are specified. \n",
    "# The function automatically recognizes the format (Currently MQ, Spectronaut, DIA-NN configured)\n",
    "input_data = aqutils.import_data(input_file)\n",
    "display(input_data)\n",
    "\n",
    "#get sample names from the imported table\n",
    "samplenames = aqutils.get_samplenames_from_input_df(input_data)\n",
    "display(samplenames)\n",
    "\n",
    "#load the samplemap dataframe (in case the user uploads a file. Basically a pandas import + separator check)\n",
    "samplemap_df = aqutils.load_samplemap(samplemap_file)\n",
    "display(samplemap_df)\n",
    "\n",
    "#compare samplemap and actual table, merge & logtransform intensities\n",
    "input_processed, samplemap_df_processed = aqutils.prepare_loaded_tables(input_data, samplemap_df)\n",
    "\n",
    "display(input_processed)\n",
    "display(samplemap_df_processed)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "import yaml\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from matplotlib_venn import venn2\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def compare_generic_table_with_original(preprocessed_input_df, original_input_file, config_yaml,input_typename_config, sep = \"\\t\"):\n",
    "    id2quant_orig, id2quant_preproc = get_processed_original_id2quant_maps(preprocessed_input_df, original_input_file, config_yaml,input_typename_config)\n",
    "    keys_orig = set(id2quant_orig.keys())\n",
    "    keys_preproc = set(id2quant_preproc.keys())\n",
    "    keydiff = keys_preproc.difference(keys_orig)\n",
    "    keys_orig = sorted(keys_orig)\n",
    "    keys_preproc = sorted(keys_preproc)\n",
    "    print(list(keys_orig)[:10])\n",
    "    print(list(keys_preproc)[:10])\n",
    "    \n",
    "    \n",
    "    assert(len(keydiff)==0) #check that all keys in the preprocessed set are part of the original set\n",
    "\n",
    "    venn2([set(id2quant_orig.keys()), set(id2quant_preproc.keys())], [\"original\", \"preprocessed\"])\n",
    "    \n",
    "    quantvec_orig = np.array([id2quant_orig.get(x)for x in id2quant_preproc.keys()])\n",
    "    quantvec_preproc = np.array([id2quant_preproc.get(x)for x in id2quant_preproc.keys()])\n",
    "    unequal_quant = [id2quant_orig.get(x)!=id2quant_preproc.get(x) for x in id2quant_preproc.keys()]\n",
    "    unequal_quant_scaled = [id2quant_orig.get(x)*10000!=id2quant_preproc.get(x) for x in id2quant_preproc.keys()]\n",
    "    print(sum(unequal_quant))\n",
    "    print(sum(unequal_quant_scaled))\n",
    "    plt.show()\n",
    "    plt.scatter(quantvec_orig, quantvec_preproc)\n",
    "    plt.show()\n",
    "    corrcoeff = np.corrcoef(quantvec_orig,quantvec_preproc)[0][1]\n",
    "    print(f\"correlation between both processings: {corrcoeff}\")\n",
    "    assert(corrcoeff>0.999)\n",
    "\n",
    "\n",
    "\n",
    "def get_processed_original_id2quant_maps(preprocessed_input_df, original_input_file, config_yaml,input_typename_config, sep = \"\\t\"):\n",
    "    config_all = yaml.safe_load(open(config_yaml, 'r'))\n",
    "    config_dict = config_all.get(input_typename_config)\n",
    "    id_cols = config_dict.get(\"ion_cols\") + [config_dict.get(\"sample_ID\")]\n",
    "    quant_col = list(config_dict.get(\"quant_ID\").values())\n",
    "    id2quant_orig = get_id2quant_original(original_input_file, id_cols, quant_col, sep)\n",
    "    id2quant_preproc = get_id2quant_processed(preprocessed_input_df, id_cols, quant_col)\n",
    "    return id2quant_orig, id2quant_preproc\n",
    "\n",
    "\n",
    "def get_id2quant_original(original_input_file, id_cols, quant_col, sep):\n",
    "    print(id_cols)\n",
    "    print(id_cols+quant_col)\n",
    "    orig_df = pd.read_csv(original_input_file, sep=sep, usecols= id_cols+quant_col)\n",
    "    orig_df[\"compareID\"] = orig_df[id_cols].astype('string').sum(axis = 1)\n",
    "    display(orig_df)\n",
    "    id2quant = dict(zip(orig_df[\"compareID\"], orig_df[quant_col[0]]))\n",
    "    id2quant = {k: round(v,3) for k, v in id2quant.items()}\n",
    "    return id2quant\n",
    "\n",
    "\n",
    "def get_id2quant_processed(preprocessed_input_df, id_cols, quant_col):\n",
    "    compare_IDs = []\n",
    "    quantvals = []\n",
    "    for column in preprocessed_input_df.columns:\n",
    "        if(column == \"protein\"):\n",
    "            continue\n",
    "        id = pd.Series([column for x in range(len(preprocessed_input_df.index))]).to_numpy()[0]\n",
    "        reformated_pep_id = [x.split(\"_MOD_\")[1].replace(\"_CHARGE_\", \"\")[:-1]+id for x in preprocessed_input_df.index]\n",
    "        compare_IDs.extend(reformated_pep_id)\n",
    "\n",
    "        quantvals.extend(list(preprocessed_input_df[column]))\n",
    "    \n",
    "    id2quant = dict(zip(compare_IDs, quantvals))\n",
    "    id2quant = {k: round(2**v,3) for k, v in id2quant.items() if ~np.isnan(v)}\n",
    "    return id2quant\n",
    "\n",
    "\n",
    "import alphaquant.diffquant_utils as aqutils\n",
    "import os\n",
    "\n",
    "\n",
    "input_file = os.path.join(tabledir, \"spectronaut.tsv\")\n",
    "samplemap_file = os.path.join(tabledir, \"samplemap.spectronaut.tsv\")\n",
    "\n",
    "\n",
    "input_data = aqutils.import_data(input_file)\n",
    "samplemap_df = aqutils.load_samplemap(samplemap_file)\n",
    "input_processed, samplemap_df_processed = aqutils.prepare_loaded_tables(input_data, samplemap_df)\n",
    "compare_generic_table_with_original(input_processed, input_file, os.path.join(\"alphaquant\",\"configs\",\"intable_config.yaml\"), \"spectronaut_precursor_v2\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.5 64-bit ('base': conda)",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
